share/spack/spack-completion.bash

```diff
@@ -91,14 +102,110 @@ _bash_completion_spack () {
     fi
 }
 
+# Helper functions for subcommands
+
+_subcommands () {
+    spack commands
+}
+
+_all_packages () {
+    spack list
+}
+
+_all_resource_hashes () {
+    spack resource list --only-hashes
+}
+
+_installed_packages () {
+    spack --color=never find --no-groups
+}
+
+_installed_compilers () {
+    spack compilers | egrep -v "^(-|=)"
+}
+
+_providers () {
+    spack providers
+}
+
+_mirrors () {
+    spack mirror list | awk '{print $1}'
+}
+
+_repos () {
+    spack repo list | awk '{print $1}'
+}
+
+_tests () {
+    spack test -l
+}
+
+_environments () {
+    spack env list
+}
+
+_keys () {
+    spack gpg list
+}
+
+_config_sections () {
+    echo "compilers mirrors repos packages modules config upstreams"
+}
+
+_extensions () {
+    echo "aspell go-bootstrap go icedtea jdk kim-api lua matlab mofem-cephas octave openjdk perl python r ruby rust tcl yorick"
```

Comment:
If we could implement #1356, that would be great. Then this function will just run `spack extensions` to get a list of `extendable = True` packages.

Reply:
See #14473